### PR TITLE
A proposition for solving Eternity2 using SeaPearl

### PR DESCRIPTION
**Pull request by tomsander1998**
This is my proposition for the implementation for solving Eternity 2 using SeaPearl. One major challenge of implementing is the high number of constraints, and the choice of which constraints are branchable. I added some data for example, along with an IOmanager. 

Important notice: I have changed a little bit the code of the AllDifferent constraint so that it can take any Array as argument. It is still not merged to Master, but one can find the code on the SeaPearl/alldiff/Matrix branch.

For now, the code only works for small instances. Possible improvement are:
- Instead of adding constraints on the equality of piece's neighbors edges, one could indicate to SeaPearl that they are necessary the same variables (similarly to what is done in MiniCP)
- Chose better heuristiques for branching
- Change the modeling

### CODE REVIEW

**Review by kimriouxparadis — APPROVED** [review]
Nice job!

**Comment by tomsander1998** [comment]
I added some nice plot, and offered 3 different modeling. Model 2 is the fastest.

**Comment by 3rdCore** [comment]
Really nice work !

File: classic_cp/eternity2/data/eternity_A [hunk 1]
```diff
@@ -0,0 +1,17 @@
+4 4
```
Comment (qcappart) [inline classic_cp/eternity2/data/eternity_A:1]: This instance (if it is the infeasible one can be removed)

File: classic_cp/eternity2/data/eternity_B [hunk 2]
```diff
@@ -0,0 +1,50 @@
+7 7
```
Comment (qcappart) [inline classic_cp/eternity2/data/eternity_B:1]: Same instance as 7x7 (and can then be removed)

File: classic_cp/eternity2/data/eternity_C [hunk 3]
```diff
@@ -0,0 +1,65 @@
+8 8
```
Comment (qcappart) [inline classic_cp/eternity2/data/eternity_C:1]: Same instance as 8x8 (and can then be removed)

**Comment by qcappart** [comment]
Good for me ! :-) Just try to find other names than v2, v3 name as it is confusing to know what are the different models.

**Review by qcappart — APPROVED** [review]

**Comment by tomsander1998** [comment]
Okay !

**Pull request merged**